### ISSUE TITLE
Support timm schedulers and other custom schedulers

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#10190]
## 🚀 Feature

Previous discussions:
https://github.com/PyTorchLightning/pytorch-lightning/issues/9964
https://github.com/PyTorchLightning/pytorch-lightning/issues/5555

### Motivation

[timm](https://github.com/rwightman/pytorch-image-models) seems to be a very popular package (14k stars). Lightning currently raises an error when using it and doesn't support it out of the box.

### Pitch
Since it has all the basic interface that we need inside lightning, we can add a hook `LightningModule.scheduler_step` similar to `optimizer_step`.

This will be the default implementation
```py
class LightningModule(...):
    ...
    def scheduler_step(self, scheduler, opt_idx, step, monitor, ...):
       
        if monitor is None:
            scheduler.step()
        else:
            scheduler.step(monitor)
```

and for timm or some other schedulers one can configure and can do:
```py
class LitModel(pl.LightningModule):
    ...
    def scheduler_step(self, scheduler, opt_idx, step, monitor, ...):
        scheduler.step(epoch=step)
```

### Alternatives

Can be done using manual optimization if we remove some assertions within codebase, but doesn't seem to be a reliable solution.

### Additional context

As always open to discussion/better suggestions :)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @borda

**Comment by ananthsub** [comment]
1. Does this change anything about the typehint for schedulers in `configure_optimizers`?
2. Does this change anything about when the scheduler step in the loop is called?
3. How does scheduler_step interact with other efforts like supporting a training step that can yield losses? In that case there, is there still an optimizer index available?
4. Could you provide an example of how timm or other custom LR schedulers would work with this new interface?

**Comment by rohitgr7** [comment]
@ananthsub 
1. Yes, not I guess we accept all sorts of object. so updating it to `Any`. any better suggestions?
2. No, it will only configure how `scheduler.step` will be done.. similar to `optimizer_step` hook.. when it will be done will still be decided using `interval` and `frequency` values set inside the config by the user.
3. yes it will, in that case we will still have optimizer_indexes assigned to the corresponding scheduler. so it will be like `trainer_step(opt_idx=0) -> optimizer_step(opt_idx=0) -> scheduler_step(opt_idx=0) -> trainer_step(opt_idx=1) (yield second time) -> optimizer_step(opt_idx=1) -> scheduler_step(opt_idx=1) -> `
4. Some of the Timm schedulers don't override `_LRScheduler` and require the epoch value inside `.step()` which is not the case with PyTorch Schedulers.
so its like
```py
scheduler.step(epoch=current_epoch)
```
since we have `current_epoch` attached to `self.trainer`, it can simply be done.

**Comment by ananthsub** [comment]
> Yes, not I guess we accept all sorts of object. so updating it to Any. any better suggestions?

This can't be `Any` since Lightning relies on an object that defines `step` for execution and `state_dict` for checkpointing

**Comment by rohitgr7** [comment]
> This can't be Any since Lightning relies on an object that defines step for execution and state_dict for checkpointing

I can add some checks for it.

> FWIW, I'd make this issue about supporting custom LR schedulers that don't follow the PyTorch LR scheduler interface, since this isn't timm specific.

yeah that is mentioned in the title, but I also mentioned `timm` since it caused this issue to come up.

> What happens about other components? For example, should someone be able to bring their own dataloader too? Where do we draw the line? A risk I see is continuing down this line, Lightning could end up dictating interfaces different from PyTorch, which breaks the promise of Lightning being organized PyTorch for training.

True, I got skeptical about this when I actually started working on it and removing the checks which are specific for pytorch native schedulers. Where do we draw the line? I don't know for this one. Couldn't find a way to support this really but it has come up a few times in the issue and over slack in the past. I don't personally use timm schedulers neither I have implemented any custom one in the past and pytorch schedulers have been working fine for me so far. If we don't go further supporting these custom schedulers, I'd suggest we still keep the hook, to allow custom schedulers sub-classed with pytorch `LRScheduler`.

**Comment by LuoXin-s** [comment]
@rohitgr7 Hi, how about the progress of this problem?

**Comment by rohitgr7** [comment]
hey, @LuoXin-s this is sort of ready, but we have certain relevant conflicting points for adding its support [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/10190#issuecomment-960302689).'

I'd need approvals from @tchaton and @ananthsub before moving forward.

**Comment by tchaton** [comment]
Hey @rohitgr7, would writing down a protocol API for schedulers resolve this ?

**Comment by awaelchli** [comment]
Have we considered providing a wrapper for custom schedulers, where a reference to the trainer or other attributes could be exposed. This would be an alternative to the LightningModule hook. On the other hand, there is also a optimizer_step hook so it seems to be natural to also have one for the scheduler.

**Comment by rohitgr7** [comment]
I think the latter one seems reasonable.

**Comment by tchaton** [comment]
@awaelchli IMO, it seems quite natural to have the scheduling exposed as hook as it is key component of the optimization and all others are exposed already.

I would prefer not expose the wrapping approach as users might need information not available through the Trainer and this increase onboarding for new users.
